plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php

```diff
@@ -0,0 +1,371 @@
+				sprintf(
+					/* translators: 1: the size of the payload, 2: the maximum allowed payload size */
+					__( 'JSON payload size is %1$s bytes which is larger than the maximum allowed size of %2$s bytes.', 'optimization-detective' ),
+					number_format_i18n( $content_length ),
+					number_format_i18n( $max_size )
+				),
+				array( 'status' => 413 )
+			);
+		}
+
+		try {
+			$url_metric_group->add_url_metric( $url_metric );
+		} catch ( InvalidArgumentException $e ) { // @codeCoverageIgnore
+			// NOTE: This exception should never be thrown because `get_group_for_viewport_width()` already ensures the viewport width is valid.
+			return new WP_Error( 'invalid_url_metric', $e->getMessage() ); // @codeCoverageIgnore
+		}
+
+		$result = OD_URL_Metrics_Post_Type::update_post(
+			$request->get_param( 'slug' ),
+			$url_metric_group_collection
+		);
+		if ( $result instanceof WP_Error ) {
+			$error_data = array(
+				'status' => 500,
+			);
+			if ( WP_DEBUG ) {
+				$error_data['error_code']    = $result->get_error_code();
+				$error_data['error_message'] = $result->get_error_message();
+			}
+			return new WP_Error(
+				'unable_to_store_url_metric',
+				__( 'Unable to store URL Metric.', 'optimization-detective' ),
+				$error_data
+			);
+		}
+		$post_id = $result;
+
+		// Schedule an event in 10 minutes to trigger an invalidation of the page cache (hopefully).
+		$cache_purge_post_id = $request->get_param( 'cache_purge_post_id' );
+		if ( is_int( $cache_purge_post_id ) && false === wp_next_scheduled( 'od_trigger_page_cache_invalidation', array( $cache_purge_post_id ) ) ) {
+			wp_schedule_single_event(
+				time() + 10 * MINUTE_IN_SECONDS,
+				'od_trigger_page_cache_invalidation',
+				array( $cache_purge_post_id )
+			);
+		}
+
+		/**
+		 * Fires whenever a URL Metric was successfully stored.
+		 *
+		 * @since 0.7.0
+		 *
+		 * @param OD_URL_Metric_Store_Request_Context $context Context about the successful URL Metric collection.
+		 */
+		do_action(
+			'od_url_metric_stored',
+			new OD_URL_Metric_Store_Request_Context(
+				$request,
+				$post_id,
+				$url_metric_group_collection,
+				$url_metric_group,
+				$url_metric
+			)
+		);
+
+		return new WP_REST_Response(
+			array(
+				'success' => true,
+			)
+		);
+	}
+
+	/**
+	 * Decompresses the REST API request body for the URL Metrics endpoint.
+	 *
+	 * @since n.e.x.t
+	 * @access private
+	 *
+	 * @phpstan-param WP_REST_Request<array<string, mixed>> $request
+	 *
+	 * @param mixed           $result  Response to replace the requested version with. Can be anything a normal endpoint can return, or null to not hijack the request.
+	 * @param WP_REST_Server  $server  Server instance.
+	 * @param WP_REST_Request $request Request used to generate the response.
+	 * @return mixed Passed through $result if successful, or otherwise a WP_Error.
+	 */
+	public function decompress_rest_request_body( $result, WP_REST_Server $server, WP_REST_Request $request ) {
```

Comment:
Having this in here doesn't seem ideal, since the method is not scoped to this endpoint (even though it technically only acts if the endpoint is hit).

So I think ideally this would be either a standalone function or in a general REST utility class with static methods.

Since it's marked private, not a blocker for merging, but wanted to flag. It could be refactored later.

Reply:
OK, I'll address in a follow-up I'm working on right now.

Reply:
See https://github.com/WordPress/performance/pull/1924